### ISSUE TITLE
Repair operation uses more "non-lsa" memory since upgrade to 5.4

### ISSUE THREAD

**Issue by Skunnyk** [scylladb/scylladb#17530]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 5.4.2
Cluster size: 14 nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): Debian

*Hardware details (for performance issues)*          
Platform (physical/VM/cloud instance type/docker): physical
Hardware: sockets=1 cores=8 hyperthreading=16 memory=64GB
Disks: 2 nvme 1.9TB raid 0 / ~50% disk space usage.

We have billions of small partitions. We already reduced `bloom_filter_fp_chance` to `0.1` because the default value (`0.01`) uses too much RAM for bloom filters with LCS.

We have a weird behavior on repair on our clusters upgraded to 5.4.2 (from 5.2.x). 
Looks like too much memory is used by repair.

The "non-lsa" memory graph speak by itself: Upgrade was on the 8 February, and on the 1st 5.4.2 repair (February, 11), we can see a huge peak of Non-LSA, from ~22GB/nodes to 48GB !
So no more ram available for the engine, and tons of `bad::alloc` with  `[shard 11:stat] storage_proxy - exception during mutation write to 10.10.11.42: logalloc::bad_alloc (failed to refill emergency reserve of 30 (have 29 free segments))` . 
![Screenshot_2024-02-26_15-56-29](https://github.com/scylladb/scylladb/assets/3627062/3ac846d1-e3c8-49aa-9efb-b1c773985854)

With 5.2, repair non-lsa memory peak were ~28GB.
![Screenshot_2024-02-26_15-56-01](https://github.com/scylladb/scylladb/assets/3627062/39c9508c-6882-4fe7-92e6-4df174541d5e)

We disabled repair for now, but we can reproduce the behavior change on all our cluster, with different schemas.
It seems that during the repair, tons of sstables are created, and almost no compaction  (from ~7500 sstables on the keyspace, to 12000, at the moment of the crash/bad_alloc). 

I can find a few related issues, like https://github.com/scylladb/scylladb/issues/15665 or https://github.com/scylladb/scylladb/issues/16362 , and maybe this comment `I can conclude that the problem is with how repair accounts memory. Repair has a memory limit it respects, but the only thing it accounts for is the frozen mutation fragment held in each repair_row. It doesn't account for anything else, in particular it doesn't account for keys. But in this test, there are rows with keys of size ~8 kiB and values of size ~16 B. (These rows come from materialized views). So the accounting is completely dysfunctional in this case, and repair ends up swallowing unlimited amounts of memory. (E.g. I witnessed a std::list holding multiple GiB worth of rows in the core, even though it was supposed to be limited to 32 MiB).`

**Comment by denesb** [comment]
The problem with accounting is present in all versions, it is not new in 5.4.
Also, repair indeed writes many tiny sstables, these are compacted at the end of repair, with off-strategy compaction. Is this new in 5.4? @raphaelsc

**Comment by Skunnyk** [comment]
I can see the behavior change on all our cluster (more non-lsa ram used during repair, even on smaller ones, but without OOM because less data), I wonder how this problem has not been spotted by other people :thinking:

**Comment by denesb** [comment]
@Skunnyk can we get a coredump, taken when the issue is happening?

**Comment by Skunnyk** [comment]
I have coredumps on some nodes, dating back from the original crash! (I also have others ones when I did "repair" testings.

`xz` running on the coredump, I'll upload it based on https://opensource.docs.scylladb.com/stable/troubleshooting/report-scylla-problem.html#core-dump

**Comment by avikivity** [comment]
Please check bloom filter file sizes.

**Comment by denesb** [comment]
> @denesb : The coredump has been uploaded with uuid `c538ced6-3187-4bbb-a9bb-cf59cc61f5e2` (I had a message `<?xml version='1.0' encoding='UTF-8'?><Error><Code>AccessDenied</Code><Message>Access denied.</Message><Details>Anonymous caller does not have storage.objects.delete access to the Google Cloud Storage object.</Details></Error>` at the end of the curl upload so I don't know if it's ok… ?).

I downloaded this but I got an empty file. Try generating a new UUID, maybe you tried to overwrite an existing one?

**Comment by Skunnyk** [comment]
Uploaded again as `bccd5fed-f77b-4432-af11-15c9c66c2ad1`, no errors :crossed_fingers:

**Comment by denesb** [comment]
> Uploaded again as `bccd5fed-f77b-4432-af11-15c9c66c2ad1`, no errors 🤞

Thanks @Skunnyk, I managed to download and open the core. I will dig in tomorrow.

**Comment by denesb** [comment]
It is the bloom filters. Small excerpt from `scylla sstables` output:
```
(sstables::sstable*) 0x604083355180: local=1 data_file=4452919, in_memory=6167904 (bf=6160840, summary=6440, sm=0) "keyspace"."table" filename=me-3gdy_1efa_1hl0g2uo43gaasksv5-big-Data.db
(sstables::sstable*) 0x604091016a80: local=1 data_file=3504063, in_memory=3150960 (bf=3146000, summary=4336, sm=0) "keyspace"."table" filename=me-3gdy_1efa_3yqw02uo43gaasksv5-big-Data.db
(sstables::sstable*) 0x60407419e380: local=1 data_file=6672024, in_memory=4599204 (bf=4587880, summary=10700, sm=0) "keyspace"."table" filename=me-3gdy_1efb_02klc2uo43gaasksv5-big-Data.db
(sstables::sstable*) 0x604082e9d180: local=1 data_file=8668697, in_memory=4730400 (bf=4718960, summary=10816, sm=0) "keyspace"."table" filename=me-3gdy_1egi_0jag02uo43gaasksv5-big-Data.db
(sstables::sstable*) 0x6040a5068000: local=1 data_file=5547270, in_memory=3153128 (bf=3146000, summary=6504, sm=0) "keyspace"."table" filename=me-3gdy_1eo5_3dba82uo43gaasksv5-big-Data.db
(sstables::sstable*) 0x60406d084a80: local=1 data_file=24578671, in_memory=9339664 (bf=9306760, summary=32280, sm=0) "keyspace"."table" filename=me-3gdy_1eo6_06fhc2uo43gaasksv5-big-Data.db
(sstables::sstable*) 0x604008e56e00: local=1 data_file=722602, in_memory=2100620 (bf=2097360, summary=2636, sm=0) "keyspace"."table" filename=me-3gdy_1eqp_1q5nk2uo43gaasksv5-big-Data.db
(sstables::sstable*) 0x60405eedaa80: local=1 data_file=907071, in_memory=3149272 (bf=3146000, summary=2648, sm=0) "keyspace"."table" filename=me-3gdy_1eqp_31qm82uo43gaasksv5-big-Data.db
(sstables::sstable*) 0x60405eeda700: local=1 data_file=710976, in_memory=3280340 (bf=3277080, summary=2636, sm=0) "keyspace"."table" filename=me-3gdy_1eqq_0nkrk2uo43gaasksv5-big-Data.db
```

Many of the tiny sstables above have bloom filters that are as big or even bigger as the Data component itself. These sstables are probably produced by repair.

Summary:
```
total (shard-local): count=894, data_file=105916739349384, in_memory=2976371924
```

There are 4.4GB memory available on the shard, of which 2.9GB is taken by in-memory components of sstables.

This is https://github.com/scylladb/scylladb/issues/2024, which has been a problem for some time now. I don't know why 5.4 made this worse. @raphaelsc did we make changes around bloom filters recently?

**Comment by avikivity** [comment]
Possible cause: https://github.com/scylladb/scylladb/commit/4592bbe18207e9b25d485ba37103de9a35574a1d

**Comment by avikivity** [comment]
@Skunnyk can you test a version with that patch reverted? I can prepare it for you.

**Comment by avikivity** [comment]
> Hi,
> 
> @denesb : The coredump has been uploaded with uuid `c538ced6-3187-4bbb-a9bb-cf59cc61f5e2` (I had a message `<?xml version='1.0' encoding='UTF-8'?><Error><Code>AccessDenied</Code><Message>Access denied.</Message><Details>Anonymous caller does not have storage.objects.delete access to the Google Cloud Storage object.</Details></Error>` at the end of the curl upload so I don't know if it's ok… ?).
> 
> @avikivity : About `bloom filter file sizes`, you mean, the bloom filter of the files created _during_ the repair ? Or the current bloom filter size ?

Probably both.



> 
> Here is an extract of current files: LCS, `bloom_filter_fp_chance = 0.1`:
> 
> ```
> -rw-r--r-- 1 scylla scylla  102 Mar  7 07:29 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3ge8_0i1l_18kxs2pdrio3akfmzj-big-TOC.txt
> -rw-r--r-- 1 scylla scylla  87K Mar  7 07:30 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3ge8_0i1l_18kxs2pdrio3akfmzj-big-Summary.db
> -rw-r--r-- 1 scylla scylla  87M Mar  7 07:30 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3ge8_0i1l_18kxs2pdrio3akfmzj-big-Index.db
> -rw-r--r-- 1 scylla scylla 1.8M Mar  7 07:30 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3ge8_0i1l_18kxs2pdrio3akfmzj-big-Filter.db
> -rw-r--r-- 1 scylla scylla   10 Mar  7 07:30 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-
> ```

Filter/Data ratio = 0.02. Tight but ok-ish.

> ```
> 3ge8_0i1l_18kxs2pdrio3akfmzj-big-Digest.crc32
> -rw-r--r-- 1 scylla scylla 160M Mar  7 07:30 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3ge8_0i1l_18kxs2pdrio3akfmzj-big-Data.db
> -rw-r--r-- 1 scylla scylla 6.3K Mar  7 07:30 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3ge8_0i1l_18kxs2pdrio3akfmzj-big-Statistics.db
> -rw-r--r-- 1 scylla scylla 4.0K Mar  7 07:30 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3ge8_0i1l_18kxs2pdrio3akfmzj-big-Scylla.db
> -rw-r--r-- 1 scylla scylla 556K Mar  7 07:30 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3ge8_0i1l_18kxs2pdrio3akfmzj-big-CompressionInfo.db
> ```
> 
> Old ssttables, LCS with `bloom_filter_fp_chance = 0.01` (default value)
> 
> ```
> -rw-r--r-- 1 scylla scylla 370K Jan 18 13:32 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3163110-big-CompressionInfo.db
> -rw-r--r-- 1 scylla scylla 111M Jan 18 13:32 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3163110-big-Data.db
> -rw-r--r-- 1 scylla scylla   10 Jan 18 13:32 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3163110-big-Digest.crc32
> -rw-r--r-- 1 scylla scylla 2.3M Jan 18 13:32 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3163110-big-Filter.db
> ```

Also 0.02

> ```
> -rw-r--r-- 1 scylla scylla  68M Jan 18 13:32 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3163110-big-Index.db
> -rw-r--r-- 1 scylla scylla  347 Jan 18 13:32 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3163110-big-Scylla.db
> -rw-r--r-- 1 scylla scylla 6.3K Jan 18 13:32 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3163110-big-Statistics.db
> -rw-r--r-- 1 scylla scylla  60K Jan 18 13:32 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3163110-big-Summary.db
> -rw-r--r-- 1 scylla scylla  102 Jan 18 13:32 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3163110-big-TOC.txt
> ```
> 
> I don't have an extract during the repair, but I can run a small repair on the cluster to check, if needed.

The repair ones are likely the problem.

**Comment by Skunnyk** [comment]
Hi,

Thank you for your answer.

> @Skunnyk can you test a version with that patch reverted? I can prepare it for you.

Hm, I might be able to test in a staging env, I need to prepare it with an extract of the production. We are using `5.4.3-0.20240211.cf42ca0c2a65-1` debian version for now. Not sure if it's a good idea to try it on one node of the problematic cluster :) 


I ran a small repair in production, and yes, the filter are huge :o :
```
-rw-r--r-- 1 scylla scylla  102 Mar 11 17:40 /srv/scylla/data/keyspace/able-8c19d780e73b11e7bd40000000000002/me-3gec_1aaw_2wdps26t7oh38no9pb-big-TOC.txt
-rw-r--r-- 1 scylla scylla 905K Mar 11 17:40 /srv/scylla/data/keyspace/able-8c19d780e73b11e7bd40000000000002/me-3gec_1aaw_2wdps26t7oh38no9pb-big-Index.db
-rw-r--r-- 1 scylla scylla 1.7M Mar 11 17:40 /srv/scylla/data/keyspace/able-8c19d780e73b11e7bd40000000000002/me-3gec_1aaw_2wdps26t7oh38no9pb-big-Data.db
-rw-r--r-- 1 scylla scylla   10 Mar 11 17:40 /srv/scylla/data/keyspace/able-8c19d780e73b11e7bd40000000000002/me-3gec_1aaw_2wdps26t7oh38no9pb-big-Digest.crc32
-rw-r--r-- 1 scylla scylla  980 Mar 11 17:40 /srv/scylla/data/keyspace/able-8c19d780e73b11e7bd40000000000002/me-3gec_1aaw_2wdps26t7oh38no9pb-big-Summary.db
-rw-r--r-- 1 scylla scylla 5.9M Mar 11 17:40 /srv/scylla/data/keyspace/able-8c19d780e73b11e7bd40000000000002/me-3gec_1aaw_2wdps26t7oh38no9pb-big-Filter.db
-rw-r--r-- 1 scylla scylla 6.3K Mar 11 17:40 /srv/scylla/data/keyspace/able-8c19d780e73b11e7bd40000000000002/me-3gec_1aaw_2wdps26t7oh38no9pb-big-Statistics.db
-rw-r--r-- 1 scylla scylla 6.1K Mar 11 17:40 /srv/scylla/data/keyspace/able-8c19d780e73b11e7bd40000000000002/me-3gec_1aaw_2wdps26t7oh38no9pb-big-CompressionInfo.db
-rw-r--r-- 1 scylla scylla  343 Mar 11 17:40 /srv/scylla/data/keyspace/able-8c19d780e73b11e7bd40000000000002/me-3gec_1aaw_2wdps26t7oh38no9pb-big-Scylla.db




-rw-r--r-- 1 scylla scylla  102 Mar 11 17:39 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3gec_1aa3_5vjqo2nxc47ebj0ylb-big-TOC.txt
-rw-r--r-- 1 scylla scylla 5.2M Mar 11 17:42 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3gec_1aa3_5vjqo2nxc47ebj0ylb-big-Index.db
-rw-r--r-- 1 scylla scylla 5.3K Mar 11 17:42 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3gec_1aa3_5vjqo2nxc47ebj0ylb-big-Summary.db
-rw-r--r-- 1 scylla scylla   10 Mar 11 17:42 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3gec_1aa3_5vjqo2nxc47ebj0ylb-big-Digest.crc32
-rw-r--r-- 1 scylla scylla 9.6M Mar 11 17:42 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3gec_1aa3_5vjqo2nxc47ebj0ylb-big-Data.db
-rw-r--r-- 1 scylla scylla 6.3K Mar 11 17:42 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3gec_1aa3_5vjqo2nxc47ebj0ylb-big-Statistics.db
-rw-r--r-- 1 scylla scylla  431 Mar 11 17:42 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3gec_1aa3_5vjqo2nxc47ebj0ylb-big-Scylla.db
-rw-r--r-- 1 scylla scylla  18M Mar 11 17:42 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3gec_1aa3_5vjqo2nxc47ebj0ylb-big-Filter.db
-rw-r--r-- 1 scylla scylla  35K Mar 11 17:42 /srv/scylla/data/keyspace/table-8c19d780e73b11e7bd40000000000002/me-3gec_1aa3_5vjqo2nxc47ebj0ylb-big-CompressionInfo.db
```

Data: 9.6MB / Filter: 18M . What is happening :thinking:

**Comment by avikivity** [comment]
I don't think a version with that patch reverted poses any risk to run in production.

**Comment by avikivity** [comment]
Well, with very small partitions a relatively large bloom filter is expected. What's surprising is that it's much larger than the other sstables.

**Comment by raphaelsc** [comment]
> Well, with very small partitions a relatively large bloom filter is expected. What's surprising is that it's much larger than the other sstables.

The only explanation I have is that it's generated by a compaction that expired lots of data.

**Comment by Skunnyk** [comment]
I reproduced the problem in staging, with a repair where I had a sstable with 2.1MB of Filter and 54K of Data. So I might be able to test a patched version if needed.

**Comment by denesb** [comment]
@Skunnyk I prepared a build, based on this branch: https://github.com/denesb/scylla/tree/bloom-filter-regression-search
It reverts two bloom-filter related patches, which I suspect could be culprits here:
* https://github.com/denesb/scylla/commit/4592bbe18207e9b25d485ba37103de9a35574a1d
* https://github.com/denesb/scylla/commit/da04fea71e59ca32948d38e4de9f10bbbc66f762

You can find the packages here https://scratch.scylladb.com/scylla-5.4.2-1.bloom-filter.tar.gz

Can you please install this on your staging environment and re-run the reproducer and report back whether it still reproduces or not?

**Comment by avikivity** [comment]
@denesb with the first revert, the second is immaterial, no? Since we'll just ignore the estimation.

**Comment by denesb** [comment]
> @denesb with the first revert, the second is immaterial, no? Since we'll just ignore the estimation.

The two reverts belong to different subsystems: the first changes compaction, the second changes repair. They are orthogonal to each other.

**Comment by avikivity** [comment]
Yes, and the other one is in fact relevant to tiny partitions.

**Comment by raphaelsc** [comment]
note that even though the patches increase memory consumption for filters, they're technically correct. they're a problem for certain users (e.g. tons of tiny partitions), but required for others. the 1st patch fixes estimation for sstables produced by repair, but note we'll rewrite the sstables in off-strategy later, so the issue is only transient.

**Comment by Skunnyk** [comment]
Thank you! 

If the (reverted) patches are correct and expected, do you need me to test the provided 5.4.2-bloomfilter version ? 

For information, our schema is like:

**Comment by Skunnyk** [comment]
Thank you! 

If the (reverted) patches are correct and expected, do you need me to test the provided 5.4.2-bloomfilter version ? 
I tested it on a staging cluster, but I don't have enough data to see any significant changes I guess :( 

For information, our schema is something like that:
```
CREATE TABLE keyspace.table (
    id ascii,
    document_id ascii,
    infused_at timestamp,
    tags set<ascii>,
    PRIMARY KEY ((id, document_id))
) WITH bloom_filter_fp_chance = 0.1
  AND compaction = {'class': 'LeveledCompactionStrategy'}
  AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.DeflateCompressor'};
```

Yes, tons of partitions (billions), I don't know if this is considered tiny partitions :)

**Comment by mykaul** [comment]
> AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.DeflateCompressor'};

First time I'm seeing anyone using Deflate (instead of org.apache.cassandra.io.compress.ZstdCompressor ). I don't think it's related to the issue, though.

**Comment by mykaul** [comment]
> WITH bloom_filter_fp_chance = 0.1

Isn't the default 'bloom_filter_fp_chance = 0.01' ? I wonder if this is relevant here.

**Comment by Skunnyk** [comment]
> > AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.DeflateCompressor'};
> 
> First time I'm seeing anyone using Deflate (instead of org.apache.cassandra.io.compress.ZstdCompressor ). I don't think it's related to the issue, though.

We switched to Deflate years ago, to gain something like ~20% of disk space for a tiny portion of cpu/latency (https://www.scylladb.com/2019/10/07/compression-in-scylla-part-two/)



> > WITH bloom_filter_fp_chance = 0.1
> 
> Isn't the default 'bloom_filter_fp_chance = 0.01' ? I wonder if this is relevant here.

Yes, the default is `0.01`, we bump to `0.1` because with LCS (and amplified by our "tiny partitions" schema), Filter took too much RAM (non-lsa), related: https://github.com/scylladb/scylladb/issues/2125#issuecomment-284998659 + https://github.com/scylladb/scylladb/issues/2440

**Comment by Skunnyk** [comment]
Hi,

I finally ran the patched  version (5.4.2-1.bloomfilter.20240313.334af4ae22e9) on my staging cluster with ~50GB on each (4) nodes (Tiny part of production data imported via `nodetool refresh --load-and-stream`).

I started a `nodetool repair -pr keyspace` on the node with the patch.

I have something weird: all `Filter` files created during the repair on this nodes are empty :thinking: And hundred of sstables are created:
```
-rw-r--r-- 1 scylla scylla   16 Mar 22 11:34 /srv/scylla/data/keyspace/table-a9822650ad3e11edbc2dd503536e3009/me-3gen_0td3_1tsts2i041zr2jxp4p-big-Filter.db
-rw-r--r-- 1 scylla scylla  336 Mar 22 11:34 /srv/scylla/data/keyspace/table-a9822650ad3e11edbc2dd503536e3009/me-3gen_0td3_1tsts2i041zr2jxp4p-big-Data.db
-rw-r--r-- 1 scylla scylla   16 Mar 22 11:34 /srv/scylla/data/keyspace/table-a9822650ad3e11edbc2dd503536e3009/me-3gen_0td5_1r0io2p7xx3b2d9i0p-big-Filter.db
-rw-r--r-- 1 scylla scylla  514 Mar 22 11:34 /srv/scylla/data/keyspace/table-a9822650ad3e11edbc2dd503536e3009/me-3gen_0td5_1r0io2p7xx3b2d9i0p-big-Data.db
-rw-r--r-- 1 scylla scylla  849 Mar 22 11:34 /srv/scylla/data/keyspace/table-a9822650ad3e11edbc2dd503536e3009/me-3gen_0td5_4uokw2p7xx3b2d9i0p-big-Data.db
-rw-r--r-- 1 scylla scylla   16 Mar 22 11:34 /srv/scylla/data/keyspace/table-a9822650ad3e11edbc2dd503536e3009/me-3gen_0td5_4uokw2p7xx3b2d9i0p-big-Filter.db
-rw-r--r-- 1 scylla scylla  217 Mar 22 11:34 /srv/scylla/data/keyspace/table-a9822650ad3e11edbc2dd503536e3009/me-3gen_0td6_1ffuo2006zcuxptdvd-big-Data.db
-rw-r--r-- 1 scylla scylla   16 Mar 22 11:34 /srv/scylla/data/keyspace/table-a9822650ad3e11edbc2dd503536e3009/me-3gen_0td6_1ffuo2006zcuxptdvd-big-Filter.db
-rw-r--r-- 1 scylla scylla   16 Mar 22 11:34 /srv/scylla/data/keyspace/table-a9822650ad3e11edbc2dd503536e3009/me-3gen_0tcr_0kd0w2p7xx3b2d9i0p-big-Filter.db
-rw-r--r-- 1 scylla scylla 101K Mar 22 11:34 /srv/scylla/data/keyspace/table-a9822650ad3e11edbc2dd503536e3009/me-3gen_0tcr_0kd0w2p7xx3b2d9i0p-big-Data.db
```

I wonder if I did something wrong.

**Comment by denesb** [comment]
Possibly, one of the reverted patches on that branch fixed this exact issue (empty filters).

**Comment by Skunnyk** [comment]
Hi,

Can I help you with more debug or informations ?

Repair is still stopped on this cluster, this not ideal :)

One temporary solution I have in mind for our needs is to run repair by token ranges (we use cassandra-reaper), pause it when non-lsa start to be "too high", wait for off-strategy compactions to happen (~5-15min after the pause) to merge all "repair" sstables, then un-pause the repair…

**Comment by denesb** [comment]
Hi @Skunnyk, you plan for dealing with this sounds good.

> Hi,
> 
> Can I help you with more debug or informations ?
> 
> Repair is still stopped on this cluster, this not ideal :)

:(

> One temporary solution I have in mind for our needs is to run repair by token ranges (we use cassandra-reaper), pause it when non-lsa start to be "too high", wait for off-strategy compactions to happen (~5-15min after the pause) to merge all "repair" sstables, then un-pause the repair…

This sounds like a good workaround. We are working on a fix, see https://github.com/scylladb/scylladb/pull/18141, but this will take a bit to land due to other things ongoing in the next week.

**Comment by Skunnyk** [comment]
I guess our problem will be fixed by #18141 and #18205 on 5.4 ?  :)

**Comment by denesb** [comment]
> I guess our problem will be fixed by #18141 and #18205 on 5.4 ? :)

Yes. The next 5.4 minor release should fix your problems.

**Comment by Skunnyk** [comment]
Hi,

FYI, post `5.4.6` upgrade, our clusters are happy :) 

![Screenshot_2024-05-03_11-58-32](https://github.com/scylladb/scylladb/assets/3627062/bf9c3357-c516-4755-8bf5-f2ed0c7125b9)

You can see here:
- Ram usage with `5.4.2` until the 30/04 at 14:00
- The upgrade to `5.4.6` the 30/04 at 14:00, which drop some bloom filters due to https://github.com/scylladb/scylladb/commit/1aedc7372de857bb80ef54aa0b87b21a217d82a3 (Tons of `sstables_manager - Reclaimed 1889116 bytes of memory from SSTable components. Total memory reclaimed so far is 460388540 bytes` in the logs), 
- The repair from 3:20 PM until the 01/05 at 5:30AM
